File: InterviewPrep/Leetcode/JS/linkedLists/143ReorderList.js
// You are given the head of a singly linked-list. The list can be represented as:

// L0 → L1 → … → Ln - 1 → Ln
// Reorder the list to be on the following form:

// L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
// You may not modify the values in the list's nodes. Only nodes themselves may be changed.

 

// Example 1:


// Input: head = [1,2,3,4]
// Output: [1,4,2,3]
// Example 2:


// Input: head = [1,2,3,4,5]
// Output: [1,5,2,4,3]
 

// Constraints:

// The number of nodes in the list is in the range [1, 5 * 104].
// 1 <= Node.val <= 1000



////sol//////////////


//Approach:

// 1.get the mid value
// 2. reverse from mid.next(reverse the secondhalf);
// 3. make the next of mid as null ( to make split);
// 4.now have two lists.. one points to head and second one is to reverse list return value;
// 5. to link in zig zag way
// 6.create extra two pointers.. whoich holds the next consecutive values of two lists;
//7.now point the list1.next = list2 &&  list2.next=f1;
//8now move the pointers of list1 & list2 to next values


/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */

function reverseList(head) {
    let prev = null;
    let curr = head;
    let n;
    while (curr) {
        n = curr.next;
        curr.next = prev;
        prev = curr;
        curr = n;
    }
    return prev;
}


//function to find the mid ptr
function mid(head) {
    let slowptr = head;
    let fastptr = head;

    while (fastptr.next && fastptr.next.next) {
        slowptr = slowptr.next;
        fastptr = fastptr.next.next;
    }

    return slowptr;
}


var reorderList = function (head) {

    if (!head || !head.next) return head;

    let list1 = head;
    let midptr = mid(head);
    let list2 = reverseList(midptr);
    //make sure to call after reversing the list
    midptr.next = null;// half the list

    let f1 = null;
    let f2 = null;


    while (list1 && list2) {
        // hold the pointers of next consecutive values
        f1 = list1.next;
        f2 = list2.next;

        //now link the zigzag
        list1.next = list2;
        list2.next = f1;

        // now move the pointers of list1 & list2 to next elements
        list1 = f1;
        list2 = f2;

    }

};

File: InterviewPrep/TakeUForward/Blind75/LinkedLists/143ReorderList.cpp
You are given the head of a singly linked-list. The list can be represented as:

L0 → L1 → … → Ln - 1 → Ln
Reorder the list to be on the following form:

L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
You may not modify the values in the lists nodes. Only nodes themselves may be changed.

Input: head = [1,2,3,4]
Output: [1,4,2,3]

////////////////////////sol////////////////////




Approach: The solution to this problem can be divided into three steps:

Find the midpoint and split the linked list into two halves from there.
Now, Reverse the second half of the list.
Now, connect both lists (A and B) such that  A0 -> B0 -> A1 -> B1 -> A2 -> B2 ….



/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middle(ListNode* head) {
        ListNode* slowptr = head;
        ListNode* fastptr = head;
        while (fastptr && fastptr->next) {
            slowptr = slowptr->next;
            fastptr = fastptr->next->next;
        }
        return slowptr;
    }

    ListNode* reverse(ListNode* head) {
        ListNode* curr = head;
        ListNode* prev = NULL;
        ListNode* n;
        while (curr) {
            n = curr->next;
            curr->next = prev;
            prev = curr;
            curr = n;
        }
        return prev;
    }
    void reorderList(ListNode* head) {
        if (!head || !head->next)
            return;
        ListNode* mid = middle(head);
        ListNode* k = reverse(mid->next);
        mid->next = NULL;

        ListNode* c1 = head;
        ListNode* c2 = k;
        ListNode* f1 = NULL;
        ListNode* f2 = NULL;

        while (c1 != NULL and c2 != NULL) {
            f1 = c1->next; // hold the enxt pointers value of c1 n c2
            f2 = c2->next;

            c1->next = c2;
            c2->next = f1;

            c1 = f1;
            c2 = f2;
        }
    }
};


// middle - Tc : 0(n/2) Sc : 0(1)
// reverse - Tc : 0(n) Sc : 0(1)

Time Complexity (TC): O(n)
Space Complexity (SC): O(1)